src/closures/closure_operators.jl

```diff
@@ -0,0 +1,532 @@
+    j⁺¹ = incmod1(j, grid.Ny)
+    return T(0.5) * (F(i, j⁺¹, k, grid, args...) + F(i, j, k, grid, args...))
+end
+
+"""
+    ▶z_aaf(i, j, k, grid::Grid{T}, F, args...) where T
+
+Interpolate the function or callable object
+
+    `F(i, j, k, grid, args...)`
+
+from `aaf` to `aac`.
+"""
+function ▶z_aaf(i, j, k, grid::Grid{T}, F::Function, args...) where T
+    if k >= grid.Nz
+        return T(0.5) * F(i, j, k, grid, args...)
+    else
+        return T(0.5) * (F(i, j, k+1, grid, args...) + F(i, j, k, grid, args...))
+    end
+end
+
+# Convenience operators for "interpolating constants"
+▶x_caa(i, j, k, grid, F::Number, args...) = F
+▶x_faa(i, j, k, grid, F::Number, args...) = F
+▶y_aca(i, j, k, grid, F::Number, args...) = F
+▶y_afa(i, j, k, grid, F::Number, args...) = F
+▶z_aac(i, j, k, grid, F::Number, args...) = F
+▶z_aaf(i, j, k, grid, F::Number, args...) = F
+
+# 12 operators for 4-way interpolation.
+▶xy_ffa(i, j, k, grid, F, args...) =
+ ▶y_afa(i, j, k, grid,
+ ▶x_faa, F, args...)
+
+▶xy_cfa(i, j, k, grid, F, args...) =
+ ▶y_afa(i, j, k, grid,
+ ▶x_caa, F, args...)
+
+▶xy_cca(i, j, k, grid, F, args...) =
+ ▶y_aca(i, j, k, grid,
+ ▶x_caa, F, args...)
+
+▶xy_fca(i, j, k, grid, F, args...) =
+ ▶y_aca(i, j, k, grid,
+ ▶x_faa, F, args...)
+
+▶xz_faf(i, j, k, grid, F, args...) =
+ ▶z_aaf(i, j, k, grid,
+ ▶x_faa, F, args...)
+
+▶xz_caf(i, j, k, grid, F, args...) =
+ ▶z_aaf(i, j, k, grid,
+ ▶x_caa, F, args...)
+
+▶xz_cac(i, j, k, grid, F, args...) =
+ ▶z_aac(i, j, k, grid,
+ ▶x_caa, F, args...)
+
+▶xz_fac(i, j, k, grid, F, args...) =
+ ▶z_aac(i, j, k, grid,
+ ▶x_faa, F, args...)
+
+▶yz_aff(i, j, k, grid, F, args...) =
+ ▶z_aaf(i, j, k, grid,
+ ▶y_afa, F, args...)
+
+▶yz_acf(i, j, k, grid, F, args...) =
+ ▶z_aaf(i, j, k, grid,
+ ▶y_aca, F, args...)
+
+▶yz_acc(i, j, k, grid, F, args...) =
+ ▶z_aac(i, j, k, grid,
+ ▶y_aca, F, args...)
+
+▶yz_afc(i, j, k, grid, F, args...) =
+ ▶z_aac(i, j, k, grid,
+ ▶y_afa, F, args...)
+
+"""
+    ∂x_faa(κ, Σ, i, j, k, grid, closure, u, v, w, T, S)
+
+Compute ``\\partial_x ( \\kappa \\Sigma )``, where `κ` and `Σ`
+are diffusivity and strain functions that return quantities
+at the location `faa`, where `a` is either `f` or `c`.
+The quantity returned has the location `caa`.
+"""
+function ∂x_faa(κ::Function, Σ::Function, i, j, k, grid, closure, u, v, w, T, S)
+    i⁺¹ = incmod1(i, grid.Nx)
+
+    Σ⁺¹ = Σ(i⁺¹, j, k, grid, u, v, w)
+    Σ⁻¹ = Σ(i,   j, k, grid, u, v, w)
+
+    κ⁺¹ = κ(i⁺¹, j, k, grid, closure, u, v, w, T, S)
+    κ⁻¹ = κ(i,   j, k, grid, closure, u, v, w, T, S)
+
+    return (κ⁺¹ * Σ⁺¹ - κ⁻¹ * Σ⁻¹) / grid.Δx
+end
+
+"""
+    ∂x_caa(κ, Σ, i, j, k, grid, closure, u, v, w, T, S)
+
+Compute ``\\partial_x ( \\kappa \\Sigma )``, where `κ` and `Σ`
+are diffusivity and strain functions that return quantities
+at the location `caa`, where `a` is either `f` or `c`.
+The quantity returned has the location `faa`.
+"""
+function ∂x_caa(κ::Function, Σ::Function, i, j, k, grid, closure, u, v, w, T, S)
```

Comment:
I wonder if these functions can be defined in `@eval` loops. Might be dangerous path as then it gets tricky to debug and makes the function definitions harder to read (and incomprehensible to someone not familiar with Julia).